examples/hmm.py

```diff
@@ -0,0 +1,179 @@
+import argparse
+
+import numpy as onp
+
+import jax.numpy as np
+from jax import lax, random
+from jax.config import config as jax_config
+from jax.scipy.special import logsumexp
+
+import numpyro.distributions as dist
+from numpyro.handlers import sample
+from numpyro.hmc_util import initialize_model
+from numpyro.mcmc import hmc
+from numpyro.util import fori_collect
+
+
+"""
+In this example, we will follow [1] to construct a semi-supervised Hidden Markov
+Model for a generative model with observations are words and latent variables
+are categories. Instead of automatically marginalizing all discrete latent
+variables (as in [2]), we will use the "forward algorithm" (which exploits the
+conditional independent of a Markov model - see [3]) to iteratively do this
+marginalization.
+
+The semi-supervised problem is chosen instead of an unsupervised one because it
+is hard to make the inference works for an unsupervised model (see the
+discussion [4]). On the other hand, this example also illustrates the usage of
+JAX's `lax.scan` primitive. The primitive will greatly improve compiling and
```

Comment:
I'm curious where the benefit of using `scan` comes in - is it primarily in terms of a lower compile time, or is it also faster after compilation? My guess would be the former.

Reply:
Yes, the first is the main benefit of lax.scan. I am not sure if it improves execution time so I'll remove it.